### PR TITLE
Improvements to AutoConfigureWorkflow

### PR DESCRIPTION
**Pull request by sfriedowitz**
## Description 
The goal of this PR is to make the `AutoConfigureWorkflow` a bit more usable by identifying some low-hanging, quality of life improvements. The central focus is on asset re-usability and searching -- if something goes wrong, have the machinery in place to identify which assets were created in the workflow, and update them in future runs rather than cluttering with new registers.

This is a step in the right direction, but suffers from the flaw that asset searching is done on platform using a `name` passed to the `AutoConfigureWorkflow` class. This is more error prone than UID requests. Ideally, all assets under a given workflow would be coordinated with a single UID, and could be retrieved/updated based on that. Until that is a possibility, the name searching approach will have to suffice.

Additions:
1) All registrations are handled by `create_or_update` calls to re-use previously configured assets with the same name
2) An `update` method that refreshes the currently stored assets, status, and status_info, in case of UI-based changes that should be pulled local
3) The option to pass a score or objective. If an objective is passed, the associated GemTable is read to extract baseline information, and an LIScore is constructed by default for design execution

TODO:
1) Update docs of class to indicate the importance of the `name` argument, and how we use the name to search for re-usable assets on platform
2) Test coverage for the new additions
3) If possible, a more coherent strategy for naming/re-using PEWs, which are registered and triggered by default in `create_default` calls before they can be renamed

### CODE REVIEW

File: src/citrine/builders/descriptors.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ def from_material(
             project: Project,
             material: Union[str, UUID, LinkByUID, MaterialRun],
             mode: AutoConfigureMode = AutoConfigureMode.PLAIN,
-            full_history: bool = False
+            full_history: bool = True
```
Comment (sfriedowitz) [inline src/citrine/builders/descriptors.py:155]: This interface change was actually a minor bug fix.

File: src/citrine/builders/auto_configure.py [hunk 2]
```diff
@@ -178,21 +188,143 @@ def design_workflow(self) -> Optional[DesignWorkflow]:
...
+    def design_executions(self) -> List[DesignExecution]:
         """Get the design execution, if it was created by this object."""
-        return self._design_execution
+        dw = self.design_workflow
+        return list(dw.design_executions.list()) if dw is not None else []
+
+    @property
+    def scores(self) -> List[Score]:
+        """Get the score, if it was created by this object."""
+        return [de.score for de in self.design_executions]
```
[the remark this answers is not in the document]
Comment (maxhutch) [reply src/citrine/builders/auto_configure.py:205]: What problem is this change meant to solve?

File: src/citrine/builders/auto_configure.py [hunk 3]
```diff
@@ -562,5 +714,31 @@ def _design_workflow_build_stage(
                 raise RuntimeError("Design workflow validation failed,"
                                    " cannot trigger design execution.")
         elif score is not None:
-            print("Triggering design execution using provided score...")
-            self.execute(score=score)
+            print("Triggering design execution...")
+            return self.execute(score=score)
+
+    def _build_default_score(self, *, objective: Objective) -> Score:
```
Comment (maxhutch) [inline src/citrine/builders/auto_configure.py:720]: I think this could be its own builder method; it seems like it could be broadly useful.

**Review by maxhutch — COMMENTED** [review]
I like these changes!  A few comments inline.

File: src/citrine/builders/auto_configure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfriedowitz) [reply src/citrine/builders/auto_configure.py:205]: It is related to the asset re-usability push.

For instance, if we pull down the same DesignWorkflow between multiple runs with multiple executions triggered, we no longer maintain a 1:1 relation that was assumed before. So storing all executions seems more relevant. At which point, listing them on the fly will always return up-to-date assets

[the remark this answers is not in the document]
Comment (maxhutch) [reply src/citrine/builders/auto_configure.py]: Disconnected from the design executions, I don't know what use there would be for printing out the whole list.

What about having a single `score` property which is the most recently used score?  Similarly, there could be a `candidates` that represents the most recently generated candidates.  This takes away the execution management part of the problem in the simple case, and doesn't take away the ability to drop into the full list of executions.

**Comment by sfriedowitz** [comment]
After much trial and tribulation, I worked out and revamped how to unit test this beast of API calls.

This involves adding a series of test.utils.factories and procedurally setting the fake response for each `find_collection` and `create_or_update` call within the `AutoConfigureWorkflow`.

Flagging for real review, now that we can think about what tweaks to interface we want to include in the next merge. Likely will be some changes to the storage of score/execution/etc requested.

File: src/citrine/builders/scores.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    objectives = objectives if isinstance(objectives, Iterable) else [objectives]
+
+    header_map = {}
+    objective_values = {}
+    for objective in objectives:
+        key = objective.descriptor_key
+        objective_values[key] = []
+
+        # Descriptor key has to be mapped inexactly to fieldnames of table data
+        full_key = next(filter(lambda col: key in col, reader.fieldnames), None)
```
[the remark this answers is not in the document]
Comment (sfriedowitz) [reply src/citrine/builders/scores.py:55]: We only have access to the string descriptor key, not the actual descriptor object with units. However, matching on `f"{descriptor_key}~Mean" without the units should be sufficient to discriminate between Mean and Std

**Review by maxhutch — APPROVED** [review]
LGTM.  Let's try to clean up the testing pattern in a future PR.

**Pull request merged**